File: src/main/java/com/nred/azurum_miner/entity/EmptyMatrixItemEntity.java
package com.nred.azurum_miner.entity;

import net.minecraft.core.BlockPos;
import net.minecraft.core.BlockPos.MutableBlockPos;
import net.minecraft.core.Direction;
import net.minecraft.core.particles.ParticleTypes;
import net.minecraft.core.particles.PowerParticleOption;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.util.Mth;
import net.minecraft.util.ProblemReporter;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.EntityType;
import net.minecraft.world.entity.item.ItemEntity;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.NetherPortalBlock;
import net.minecraft.world.level.block.Portal;
import net.minecraft.world.level.levelgen.structure.BoundingBox;
import net.minecraft.world.level.storage.TagValueInput;
import net.minecraft.world.level.storage.TagValueOutput;
import net.minecraft.world.level.storage.ValueInput;
import net.minecraft.world.level.storage.ValueOutput;
import net.minecraft.world.phys.AABB;
import net.minecraft.world.phys.Vec3;
import net.neoforged.bus.api.SubscribeEvent;
import net.neoforged.fml.common.EventBusSubscriber;
import net.neoforged.neoforge.event.entity.EntityTravelToDimensionEvent;

import java.util.List;
import java.util.UUID;

import static com.nred.azurum_miner.registration.EntityTypeRegistration.EMPTY_DIMENSIONAL_MATRIX_TYPE;
import static com.nred.azurum_miner.registration.ItemRegistration.DIMENSIONAL_MATRIX;
import static com.nred.azurum_miner.registration.ItemRegistration.EMPTY_DIMENSIONAL_MATRIX;
import static net.minecraft.sounds.SoundEvents.*;
import static net.minecraft.world.level.Level.NETHER;

@EventBusSubscriber
public class EmptyMatrixItemEntity extends ItemEntity {
    private int fillLevel = 0;
    private Vec3 dest = null;

    public EmptyMatrixItemEntity(EntityType<EmptyMatrixItemEntity> type, Level level) {
        super(type, level);
    }

    private EmptyMatrixItemEntity(ValueInput input, Level level) {
        this(input, level, 1);
    }

    private EmptyMatrixItemEntity(ValueInput input, Level level, int count) {
        this(EMPTY_DIMENSIONAL_MATRIX_TYPE.get(), level);
        this.load(input);
        this.setPickUpDelay(60);
        this.setUnlimitedLifetime();
        this.getItem().setCount(count);
    }

    @SubscribeEvent
    public static void onEntityTravelToDimension(EntityTravelToDimensionEvent event) {
        if (event.getEntity() instanceof ItemEntity entity && entity.getItem().is(EMPTY_DIMENSIONAL_MATRIX.get()) && event.getDimension() == NETHER) {
            event.setCanceled(true);

            ProblemReporter reporter = new ProblemReporter.Collector();
            TagValueOutput output = TagValueOutput.createWithContext(reporter, entity.registryAccess());
            entity.save(output);
            ValueInput input = TagValueInput.create(reporter, entity.registryAccess(), output.buildResult());
            entity.remove(Entity.RemovalReason.KILLED);

            EmptyMatrixItemEntity freshEntity = new EmptyMatrixItemEntity(input, entity.level());
            entity.level().addFreshEntity(freshEntity);
            if (entity.getItem().count() > 1) {
                freshEntity = new EmptyMatrixItemEntity(input, entity.level(), entity.getItem().count() - 1);
                freshEntity.uuid = UUID.randomUUID();
                entity.level().addFreshEntity(freshEntity);
            }
        }
    }

    @Override
    protected void addAdditionalSaveData(ValueOutput output) {
        super.addAdditionalSaveData(output);
        output.putInt("matrixFillLevel", fillLevel);
        if (this.dest != null) {
            output.store("matrixDest", Vec3.CODEC, dest);
        }
    }

    @Override
    protected void readAdditionalSaveData(ValueInput input) {
        super.readAdditionalSaveData(input);
        this.fillLevel = input.getIntOr("matrixFillLevel", 0);
        this.dest = input.read("matrixDest", Vec3.CODEC).orElse(null);
    }

    @Override
    public boolean isAlwaysTicking() {
        return true;
    }

    @Override
    public void tick() {
        super.tick();
        if (level() instanceof ServerLevel level) {
            if (this.dest != null) {
                if (this.fillLevel < 4) {
                    this.lerpPositionAndRotationStep(4, this.dest.x, this.dest.y, this.dest.z, this.getRotationVector().y, this.getRotationVector().x);
                    this.needsSync = true;
                }

                if (this.fillLevel % 5 == 0) {
                    level.sendParticles(PowerParticleOption.create(ParticleTypes.DRAGON_BREATH, 1f), this.getX(), this.getY(), this.getZ(), 3, 0.0, 0.4, 0.0, 0.02);
                }
                if (this.fillLevel % 30 == 0 && level.getRandom().nextBoolean()) {
                    this.playSound(BOTTLE_FILL, level.getRandom().nextFloat() * 0.6f + 0.4f, level.getRandom().nextFloat() * 0.5f + 0.5f);
                }

                this.fillLevel += 1;
                if (this.fillLevel > 2400) {
                    level.sendParticles(ParticleTypes.EXPLOSION, this.getX(), this.getY(), this.getZ(), 2, 0.0, 0.4, 0.0, 0.05);
                    this.playSound(DRAGON_FIREBALL_EXPLODE);
                    this.playSound(PLAYER_LEVELUP);
                    level.destroyBlock(new BlockPos(Mth.floor(dest.x), Mth.floor(dest.y), Mth.floor(dest.z)), false);
                    this.spawnSprintParticle();
                    this.setItem(new ItemStack(DIMENSIONAL_MATRIX.asItem(), getItem().getCount()));
                    this.setNoGravity(false);
                    this.setDefaultPickUpDelay();
                    this.dest = null;
                }
            } else {
                if (getItem().getCount() < 64) { // Merge stacks
                    for (EmptyMatrixItemEntity itemEntity : level.getEntitiesOfClass(this.getClass(), this.getBoundingBox().inflate(0.5, 0.0, 0.5), e -> e != this && areMergable(this.getItem(), e.getItem()))) {
                        if (areMergable(this.getItem(), itemEntity.getItem()) && itemEntity.dest == null) {
                            this.setItem(merge(this.getItem(), itemEntity.getItem(), this.getItem().count() + itemEntity.getItem().count()));
                            itemEntity.discard();
                        }
                    }
                }
            }
        }
    }

    public AABB findPortalBounds(BlockPos pos, Direction.Axis axis) {
        MutableBlockPos top = pos.mutable();
        MutableBlockPos bottom = pos.mutable();

        Level level = level();
        while (level.getBlockState(top.above()).getBlock() instanceof NetherPortalBlock) {
            top = top.move(Direction.UP);
        }

        while (level.getBlockState(bottom.below()).getBlock() instanceof NetherPortalBlock) {
            bottom = bottom.move(Direction.DOWN);
        }

        while (level.getBlockState(top.relative(axis.getPositive())).getBlock() instanceof NetherPortalBlock) {
            top = top.move(axis.getPositive());
        }

        while (level.getBlockState(bottom.relative(axis.getNegative())).getBlock() instanceof NetherPortalBlock) {
            bottom = bottom.move(axis.getNegative());
        }

        return AABB.of(BoundingBox.fromCorners(top, bottom));
    }

    @Override
    public void setAsInsidePortal(Portal portal, BlockPos pos) {
        if (this.level().isClientSide() || fillLevel != 0 || dest != null || !(portal instanceof NetherPortalBlock)) return;

        Direction.Axis axis = level().getBlockState(pos).getValue(NetherPortalBlock.AXIS);
        AABB portalBounds = findPortalBounds(pos, axis);

        List<? extends EmptyMatrixItemEntity> found = level().getEntitiesOfClass(this.getClass(), portalBounds);
        if (this.getItem().count() > 1 || found.stream().filter(it -> it.getItem().count() == 1).toList().size() > 1) { // Make sure only one is in the portal
            this.addDeltaMovement(new Vec3((axis == Direction.Axis.Z) ? this.getDeltaMovement().reverse().scale(1.5).x : this.getDeltaMovement().x, 0.0, (axis == Direction.Axis.X) ? this.getDeltaMovement().reverse().scale(1.5).z : this.getDeltaMovement().z));
            level().playSeededSound(null, this.getX(), this.getY(), this.getZ(), PLAYER_TELEPORT, this.getSoundSource(), 0.5f, 0.1f, -198679135428719823L);
            this.fillLevel = 1;
            this.setDefaultPickUpDelay();
            return;
        }

        Vec3 center = Vec3.atCenterOf(pos);
        AABB bounds = this.makeBoundingBox().move(this.getX(), this.getY(), center.z);
        if (axis == Direction.Axis.Z) {

            this.dest = new Vec3(AABB.ofSize(center, bounds.getXsize(), bounds.getYsize(), bounds.getZsize()).getCenter().x(), this.getY(), this.getZ());
        } else {
            this.dest = new Vec3(this.getX(), this.getY(), AABB.ofSize(center, bounds.getXsize(), bounds.getYsize(), bounds.getZsize()).getCenter().z);
        }

        this.setNoGravity(true);
        this.setDeltaMovement(Vec3.ZERO);
    }
}

File: src/main/kotlin/com/nred/azurum_miner/entity/EmptyMatrixItemEntity.kt
package com.nred.azurum_miner.entity

import com.nred.azurum_miner.item.ModItems
import com.nred.azurum_miner.item.ModItems.EMPTY_DIMENSIONAL_MATRIX_TYPE
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.core.Vec3i
import net.minecraft.core.particles.ParticleTypes
import net.minecraft.nbt.CompoundTag
import net.minecraft.server.level.ServerLevel
import net.minecraft.sounds.SoundEvents.*
import net.minecraft.util.Mth
import net.minecraft.world.entity.EntityType
import net.minecraft.world.entity.item.ItemEntity
import net.minecraft.world.item.ItemStack
import net.minecraft.world.level.Level
import net.minecraft.world.level.block.NetherPortalBlock
import net.minecraft.world.level.block.Portal
import net.minecraft.world.level.levelgen.structure.BoundingBox
import net.minecraft.world.phys.AABB
import net.minecraft.world.phys.Vec3
import kotlin.random.Random


// TODO kotlinforforge 3.9 functions not working!
fun Vec3.toVec3i(): Vec3i = Vec3i(Mth.floor(x), Mth.floor(y), Mth.floor(z))
fun BlockPos.toVec3(): Vec3 = Vec3.atLowerCornerOf(this)

class EmptyMatrixItemEntity : ItemEntity {
    var fill = 0
    var dest: Vec3? = null

    constructor(entityType: EntityType<EmptyMatrixItemEntity>, level: Level) : super(entityType, level)
    constructor(tag: CompoundTag, level: Level) : this(tag, level, 1)
    constructor(tag: CompoundTag, level: Level, count: Int) : this(EMPTY_DIMENSIONAL_MATRIX_TYPE.get(), level) {
        this.load(tag)
        this.setPickUpDelay(60)
        this.setUnlimitedLifetime()
        this.item.count = count
    }

    override fun getType(): EntityType<*> {
        return EMPTY_DIMENSIONAL_MATRIX_TYPE.get()
    }

    override fun addAdditionalSaveData(compound: CompoundTag) {
        super.addAdditionalSaveData(compound)

        compound.putInt("matrixFill", fill)
        if (this.dest != null)
            compound.put("matrixDest", this.newDoubleList(dest!!.x, dest!!.y, dest!!.z))
    }

    override fun readAdditionalSaveData(compound: CompoundTag) {
        super.readAdditionalSaveData(compound)

        fill = compound.getInt("matrixFill")
        val list = compound.getList("matrixDest", 6)
        if (list.isNotEmpty())
            dest = Vec3(list.getDouble(0), list.getDouble(1), list.getDouble(2))
    }

    override fun tick() {
        super.tick()
        if (this.dest != null) {
            if (this.fill < 4) {
                this.lerpPositionAndRotationStep(4, this.dest!!.x, this.dest!!.y, this.dest!!.z, this.rotationVector.y.toDouble(), this.rotationVector.x.toDouble())
                this.hasImpulse = true
            }

            if (this.fill % 5 == 0) {
                (level() as ServerLevel).sendParticles(ParticleTypes.DRAGON_BREATH, this.x, this.y, this.z, 3, 0.0, 0.4, 0.0, 0.02)
            }
            if (this.fill % 30 == 0 && Random.nextBoolean()) {
                this.playSound(BOTTLE_FILL, Random.nextFloat() * 0.6f + 0.4f, Random.nextFloat() * 0.5f + 0.5f)
            }


            this.fill += 1
            if (this.fill > 2400) {
                (level() as ServerLevel).sendParticles(ParticleTypes.EXPLOSION, this.x, this.y, this.z, 2, 0.0, 0.4, 0.0, 0.05)
                this.playSound(DRAGON_FIREBALL_EXPLODE)
                this.playSound(PLAYER_LEVELUP)
                level().destroyBlock(BlockPos(dest!!.toVec3i()), false)
                this.spawnSprintParticle()
                this.item = ItemStack(ModItems.DIMENSIONAL_MATRIX.asItem(), item.count)
                this.isNoGravity = false
                this.setDefaultPickUpDelay()
                this.dest = null
            }
        } else if (!level().isClientSide) {
            if (this.item.count < 64) { // Merge stacks
                for (itemEntity in this.level().getEntitiesOfClass(this::class.java, this.boundingBox.inflate(0.5, 0.0, 0.5), { e -> e != this && areMergable(this.item, e.item) })) {
                    if (areMergable(this.item, itemEntity.item) && itemEntity.dest == null) {
                        this.item = merge(this.item, itemEntity.item, this.item.count + itemEntity.item.count)
                        itemEntity.discard()
                    }
                }
            }
        }
    }

    fun findPortalBounds(pos: BlockPos, axis: Direction.Axis): AABB {
        var top = pos.toVec3().toVec3i()
        var bottom = pos.toVec3().toVec3i()

        while (level().getBlockState(BlockPos(top.above())).block is NetherPortalBlock) {
            top = top.above()
        }

        while (level().getBlockState(BlockPos(bottom.below())).block is NetherPortalBlock) {
            bottom = bottom.below()
        }

        var temp = top
        do {
            top = temp
            temp = top.relative(axis, 1)
        } while (level().getBlockState(BlockPos(temp)).block is NetherPortalBlock)

        temp = bottom
        do {
            bottom = temp
            temp = bottom.relative(axis, -1)
        } while (level().getBlockState(BlockPos(temp)).block is NetherPortalBlock)

        return AABB.of(BoundingBox.fromCorners(top, bottom))
    }

    override fun setAsInsidePortal(portal: Portal, pos: BlockPos) {
        if (level().isClientSide || fill != 0 || portal !is NetherPortalBlock) return

        val axis = level().getBlockState(pos).getValue(portal.stateDefinition.getProperty("axis")) as Direction.Axis
        val portalBounds = findPortalBounds(pos, axis)
        val found = level().getEntitiesOfClass(this::class.java, portalBounds)

        if (this.item.count > 1 || found.filter { it.item.count == 1 }.size > 1) { // Make sure only one is in the portal
            this.deltaMovement = Vec3(if (axis == Direction.Axis.Z) this.deltaMovement.reverse().scale(0.5).x else this.deltaMovement.x, 0.0, if (axis == Direction.Axis.X) this.deltaMovement.reverse().scale(0.5).z else this.deltaMovement.z)
            level().playSeededSound(null, this.x, this.y, this.z, PLAYER_TELEPORT, this.soundSource, 0.5f, 0.1f, -198679135428719823)
            this.fill = 1
            this.setDefaultPickUpDelay()
            return
        }

        val bounds = this.makeBoundingBox().move(this.x, this.y, pos.center.z)
        dest = if (axis == Direction.Axis.Z)
            Vec3(AABB.ofSize(pos.center, bounds.xsize, bounds.ysize, bounds.zsize).center.x(), this.y, this.z)
        else
            Vec3(this.x, this.y, AABB.ofSize(pos.center, bounds.xsize, bounds.ysize, bounds.zsize).center.z)

        this.isNoGravity = true
        this.deltaMovement = Vec3.ZERO
    }
}